lib/ldap_fluff/netiq_member_service.rb

```diff
@@ -0,0 +1,81 @@
+require 'net/ldap'
+
+# handles the naughty bits of posix ldap
+class LdapFluff::NetIQ::MemberService < LdapFluff::GenericMemberService
+
+  def initialize(ldap, config)
+    @attr_login = (config.attr_login || 'memberuid')
+    super
+  end
+
+  def find_user(uid, base_dn = @base)
+    user = @ldap.search(:filter => name_filter(uid), :base => base_dn)
+    raise UIDNotFoundException if (user.nil? || user.empty?)
+    user
+  end
+
+  def find_by_dn(search_dn)
+    entry, base = search_dn.split(/(?<!\\),/, 2)
+    _entry_attr, entry_value = entry.split('=', 2)
+    entry_value = entry_value.gsub('\,', ',')
+    user = @ldap.search(:filter => name_filter(entry_value, "workforceid"), :base => base)
+    raise self.class::UIDNotFoundException if (user.nil? || user.empty?)
+    user
+  end
+
+  def get_logins(userlist)
+    userlist.map(&:downcase!)
+    result = []
+
+    userlist.each do |current_user|
+       result << find_by_dn(current_user)[0][@attr_login][0]
+    end
+
+    result
+  end
+
+  # return an ldap user with groups attached
+  # note : this method is not particularly fast for large ldap systems
+  def find_user_groups(uid)
+    groups = []
+
+    @ldap.search(:filter => Net::LDAP::Filter.eq('memberuid', uid), :base => @group_base).each do |entry|
+      groups << entry[:cn][0]
+    end
+
+    user = find_user(uid)[0][:dn][0]
+    @ldap.search(:filter => Net::LDAP::Filter.eq('member', user), :base => @group_base).each do |entry|
+      groups << entry[:cn][0]
+    end
+
+    groups
+  end
+
+  def times_in_groups(uid, gids, all)
+    filters = []
+    gids.each do |cn|
+      filters << group_filter(cn)
+    end
+    group_filters = merge_filters(filters, all)
+    filter        = name_filter(uid) & group_filters
+    @ldap.search(:base => @group_base, :filter => filter).size
+  end
+
+  # AND or OR all of the filters together
+  def merge_filters(filters = [], all = false)
+    if !filters.nil? && filters.size >= 1
+      filter = filters[0]
+      filters[1..(filters.size - 1)].each do |gfilter|
+        filter = (all ? filter & gfilter : filter | gfilter)
+      end
+      return filter
+    end
```

Comment:
I found this a little bit hard to follow, but couldn't it be shortened to something like this?
```suggestion
    (filters || []).reduce(all ? :& : :|)
```

Reply:
The code is taken from `posix_member_service.rb` verbatim, so if this should be changed, we should do it for both.